pynicotine/userlist.py

```diff
@@ -26,26 +28,47 @@ def __init__(self, core, config, queue, ui_callback):
         self.core = core
         self.config = config
         self.queue = queue
+        self.buddies = {}
         self.ui_callback = None
 
         if hasattr(ui_callback, "userlist"):
             self.ui_callback = ui_callback.userlist
 
     def server_login(self):
-
-        for row in self.config.sections["server"]["userlist"]:
-            if row and isinstance(row, list):
-                user = str(row[0])
-                self.core.watch_user(user)
+        for user in self.buddies:
+            self.core.watch_user(user)
 
     def server_disconnect(self):
+
+        for user in self.buddies:
+            self.set_user_last_seen(user, online=False)
+
         if self.ui_callback:
             self.ui_callback.server_disconnect()
 
-    def add_user(self, user):
+    def load_users(self):
+
+        for row in self.config.sections["server"]["userlist"]:
+            user = row[0]
+
+            if isinstance(user, str):
+                self.add_user(user, row)
+
+    def add_user(self, user, row=None):
+
+        if user in self.buddies:
+            return
+
+        new_user = (row is None)
+
+        if not row or not isinstance(row, list):
+            row = [user, "", False, False, False, "Never seen", ""]
+
+        self.buddies[user] = row
+        self.save_user_list()
```

Comment:
During load at startup, there is a moment where the config setting is re-populated with an incomplete buddy list.